### ISSUE TITLE
window-specific attributes like width_increment etc. need to be moved from Con to Window struct

### ISSUE THREAD

**Issue by stapelberg** [i3/i3#665]
Probably due to confusion they are saved in the Con instead of in the Window. However, as the window can be re-attached to another Con (for example when using sticky windows and/or swallowing), they need to be stored in the Window.

**Comment by Airblader** [comment]
Can we identify which properties need to be moved? I think these are pretty safe:

* `base_width`
* `base_height`
* `width_increment`
* `height_increment`

What about `mapped` and `depth`?

**Comment by stapelberg** [comment]
I agree that `base_width`, `base_height`, `width_increment` and `height_increment` should be moved.

`border_width` is also set per X11 window, and `aspect_ratio` comes from the window hints, so both should be moved.

`depth` is the depth of the container window, so I currently don’t think it should be moved.

`mapped` is a corner-case, which cannot easily be moved, I think. See e.g. `tree_render()`, which changes the root node’s mapped property.

**Comment by Airblader** [comment]
What about border_style?

**Comment by stapelberg** [comment]
`border_style` should remain in the Con. As you mentioned, it’s a per-container attribute also used for tabbed/stacked containers.

**Comment by Airblader** [comment]
Yeah, but doesn't the same apply to border_width then?

**Comment by stapelberg** [comment]
I haven’t had a very close look, but I think it only applies to `current_border_width`, and `border_width` really is just determined by the X11 properties.

**Comment by Airblader** [comment]
As a note, with #1966 in mind, once that is implemented we actually should move `depth` to `Window` since containers themselves no longer have any reason to have a window. That is assuming we *never* use the parent for decoration rendering anymore, including tabbed/stacked, of course.
